### PR TITLE
Present new binary for checking hashes

### LINKED ISSUE #2348: [Feature Request] Present new Check / Get Hashes bin
It would be awesome if bucket owners / collaborators could have new bin for checking / recalculating hashes for manifest.
It's really pain to download each binary manually and run something like: `Get-hash ... .ToLower() | clip` for each file.

There are bins which are really handy ([checkver](https://github.com/lukesampson/scoop/blob/master/bin/checkver.ps1), ...) and are easily transferable into other buckets.
Could you add new lib inside scoop for hashes check?. Basically it would download all urls for each architecture, calculate hashes and edit file (like checkver do with: `checkver <App> -u`)

If there is already option please let me know.

PS: Why to use `resolve-path (split-path (split-path (scoop which scoop)))` in:

https://github.com/lukesampson/scoop-extras/blob/7112051787c42b27de906803acfbc2cf3bba86c0/bin/checkver.ps1#L1

(and other bins) when you can use `resolve-path (scoop prefix scoop)`?

### PR DESCRIPTION
**Pull request by Ash258**
- Closes #2348

Should be working, but maybe I missed some use-case or something else.

For external buckets use: something like [this](https://github.com/Ash258/scoop-Ash258/blob/master/bin/checkhashes.ps1)

### CODE REVIEW

File: bin/checkhashes.ps1 [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+    Write-Host ($message -join "`r`n") -ForegroundColor Red
+}
+
+$MANIFESTS = @()
+foreach ($single in Get-ChildItem $Dir "$App.json") {
+    $name = (strip_ext $single.Name)
+    $manifest = parse_json "$Dir\$($single.Name)"
+
+    # Skip nighly manifests, since their hash validation is skipped
+    if (($manifest.version -ne 'nightly')) {
```
Comment (r15ch13) [inline bin/checkhashes.ps1]:
```suggestion
    if ($manifest.version -eq 'nightly') {
        continue
    }
```

**Comment by r15ch13** [comment]
Made some changes to the output and add `get_hash()` function:

```diff
diff --git a/bin/checkhashes.ps1 b/bin/checkhashes.ps1
index cd5c7cb8e..31ae4111b 100644
--- a/bin/checkhashes.ps1
+++ b/bin/checkhashes.ps1
@@ -65,36 +65,38 @@ foreach ($single in Get-ChildItem $Dir "$App.json") {
     $manifest = parse_json "$Dir\$($single.Name)"
 
     # Skip nighly manifests, since their hash validation is skipped
-    if (($manifest.version -ne 'nightly')) {
-        $urls = @()
-        $hashes = @()
-
-        if ($manifest.architecture) {
-            # First handle 64bit
-            url $manifest '64bit' | ForEach-Object { $urls += $_ }
-            hash $manifest '64bit' | ForEach-Object { $hashes += $_ }
-            url $manifest '32bit' | ForEach-Object { $urls += $_ }
-            hash $manifest '32bit' | ForEach-Object { $hashes += $_ }
-        } elseif ($manifest.url) {
-            $manifest.url | ForEach-Object { $urls += $_ }
-            $manifest.hash | ForEach-Object { $hashes += $_ }
-        } else {
-            err $name 'Manifest does not contain URL property.'
-            continue
-        }
+    if ($manifest.version -eq 'nightly') {
+        continue
+    }
 
-        # Number of URLS and Hashes is different
-        if ($urls.Length -ne $hashes.Length) {
-            err $name 'URLS and hashes count mismatch.'
-            continue
-        }
+    $urls = @()
+    $hashes = @()
+
+    if ($manifest.architecture) {
+        # First handle 64bit
+        url $manifest '64bit' | ForEach-Object { $urls += $_ }
+        hash $manifest '64bit' | ForEach-Object { $hashes += $_ }
+        url $manifest '32bit' | ForEach-Object { $urls += $_ }
+        hash $manifest '32bit' | ForEach-Object { $hashes += $_ }
+    } elseif ($manifest.url) {
+        $manifest.url | ForEach-Object { $urls += $_ }
+        $manifest.hash | ForEach-Object { $hashes += $_ }
+    } else {
+        err $name 'Manifest does not contain URL property.'
+        continue
+    }
 
-        $MANIFESTS += New-Object psobject @{
-            app      = $name
-            manifest = $manifest
-            urls     = $urls
-            hashes   = $hashes
-        }
+    # Number of URLS and Hashes is different
+    if ($urls.Length -ne $hashes.Length) {
+        err $name 'URLS and hashes count mismatch.'
+        continue
+    }
+
+    $MANIFESTS += New-Object psobject @{
+        app      = $name
+        manifest = $manifest
+        urls     = $urls
+        hashes   = $hashes
     }
 }
 
@@ -114,27 +116,16 @@ $MANIFESTS | ForEach-Object {
     $actuals = @()
 
     $current.urls | ForEach-Object {
-        $expected_hash = $current.hashes[$count]
-        $algorithm = 'sha256'
-        $name = $current.app
+        $algorithm, $expected = get_hash $current.hashes[$count]
         $version = 'HASH_CHECK'
-        $tmp = $expected_hash -split ':'
-
-        if ($tmp.Length -eq 2) {
-            $algorithm = $tmp[0]
-            $expected_hash = $tmp[1]
-        }
 
-        dl_with_cache $name $version $_ $null $null -use_cache:$UseCache
+        dl_with_cache $current.app $version $_ $null $null -use_cache:$UseCache
 
-        $to_check = fullpath (cache_path $name $version $_)
+        $to_check = fullpath (cache_path $current.app $version $_)
         $actual_hash = compute_hash $to_check $algorithm
         $actuals += $actual_hash
-        if ($actual_hash -ne $expected_hash) {
+        if ($actual_hash -ne $expected) {
             $mismatched += $count
-            Write-Host 'Wrong' -ForegroundColor Red
-        } else {
-            if (!$SkipCorrect) { Write-Host 'OK' -ForegroundColor Green }
         }
         $count++
     }
@@ -145,11 +136,16 @@ $MANIFESTS | ForEach-Object {
             Write-Host 'OK' -ForegroundColor Green
         }
     } else {
-        Write-Host "$($current.app): "
+        Write-Host "$($current.app): " -NoNewline
+        Write-Host 'Mismatch found ' -ForegroundColor Red
         $mismatched | ForEach-Object {
-            Write-Host "`t$($current.urls[$_])" -ForegroundColor Red
-            Write-Host "`t`tExp:" $current.hashes[$_] -ForegroundColor Green
-            Write-Host "`t`tAct:" $actuals[$_] -ForegroundColor Red
+            $file = fullpath (cache_path $current.app $version $current.urls[$_])
+            Write-Host  "`tURL:`t`t$($current.urls[$_])"
+            if(Test-Path $file) {
+                Write-Host  "`tFirst bytes:`t$((get_magic_bytes_pretty $file ' ').ToUpper())"
+            }
+            Write-Host  "`tExpected:`t$($current.urls[$_])" -ForegroundColor Green
+            Write-Host  "`tActual:`t`t$($actuals[$_])" -ForegroundColor Red
         }
     }
 
diff --git a/lib/core.ps1 b/lib/core.ps1
index d2c2f1aff..8c63e224d 100644
--- a/lib/core.ps1
+++ b/lib/core.ps1
@@ -657,6 +657,19 @@ function format_hash_aria2([String] $hash) {
     return $hash
 }
 
+function get_hash([String] $multihash) {
+    $type, $hash = $multihash.split(':')
+    if(!$hash) {
+        # no type specified, assume sha256
+        $type, $hash = 'sha256', $multihash
+    }
+
+    if(@('md5','sha1','sha256', 'sha512') -notcontains $type) {
+        return $false, "Hash type '$type' isn't supported."
+    }
+    return $type, $hash
+}
+
 function handle_special_urls($url)
 {
     # FossHub.com
diff --git a/lib/install.ps1 b/lib/install.ps1
index 845a5f776..c446a1c06 100644
--- a/lib/install.ps1
+++ b/lib/install.ps1
@@ -672,17 +672,13 @@ function check_hash($file, $hash, $app_name) {
     Write-Host "Checking hash of " -NoNewline
     Write-Host $(url_remote_filename $url) -f Cyan -NoNewline
     Write-Host " ... " -nonewline
-    $type, $expected = $hash.split(':')
-    if(!$expected) {
-        # no type specified, assume sha256
-        $type, $expected = 'sha256', $type
-    }
 
-    if(@('md5','sha1','sha256', 'sha512') -notcontains $type) {
-        return $false, "Hash type '$type' isn't supported."
+    $algorithm, $expected = get_hash $hash
+    if($algorithm -eq $false) {
+        return $false, "Hash type '$algorithm' isn't supported."
     }
 
-    $actual = (compute_hash $file $type).ToLower()
+    $actual = (compute_hash $file $algorithm).ToLower()
     $expected = $expected.ToLower()
 
     if($actual -ne $expected) {
```

**Comment by Ash258** [comment]
Made some changes to your suggestion:

```
    if(@('md5','sha1','sha256', 'sha512') -notcontains $type) {
        return $false, "Hash type '$type' isn't supported."
    }
```

I rather return $null instead of $false. As $false implies, it's boolean always, but it's only in one use case.

Also fixed my bug, when I striped algorithm type, when there was any other than sha256, and when you force update, it just got saved without type.

**Comment by Ash258** [comment]
Should be ready to merge 🎉

**Pull request merged**